stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/message/MessageSendingService.kt

```diff
@@ -126,29 +205,110 @@ internal class MessageSendingService(
             }
             )
         enqueueAttachmentUpload(newMessage)
-        return Result.success(newMessage)
+        jobsMap[newMessage.id]?.join()
+        return if (allAttachmentsUploaded) {
+            Result.success(messageToBeSent.copy(type = Message.TYPE_REGULAR))
+        } else Result.error(ChatError("Could not upload attachments, not sending message with id ${newMessage.id}"))
     }
 
+    /**
+     * Cancels ephemeral Message.
+     * Removes message from the offline storage and memory and notifies about update.
+     *
+     * @param message [Message] to be cancelled.
+     */
+    private suspend fun cancelEphemeralMessage(message: Message): Result<Boolean> {
+        require(message.isEphemeral()) { "Only ephemeral message can be canceled" }
+        repos.deleteChannelMessage(message)
+        channel.removeLocalMessage(message)
+        return Result(true)
+    }
+
+    /**
+     * Enqueues attachment upload work.
+     */
     private fun enqueueAttachmentUpload(message: Message) {
-        uploadAttachmentsWorker.enqueueJob(
-            channelController.channelType,
-            channelController.channelId,
-            message.id
-        )
+        uploadAttachmentsWorker.enqueueJob(channelType, channelId, message.id)
     }
 
+    // TODO: Remove this method when ChatDomain is removed.
     private suspend fun doSend(message: Message): Result<Message> {
-        val messageToSend = message.copy(type = "regular")
-        return Result.success(messageToSend)
+        return Result.success(message)
```

Comment:
There's no need for `Result.success` here, we use can imperative code, instead of a monadic APi: 
```
private suspend fun doSend(message: Message): Result<Message> {
    val channelClient = chatClient.channel(message.cid)
    val call = channelClient.sendMessageInternal(message)

    return call.retry(scope, chatClient.retryPolicy)
        .await()
        .mapSuspend(::handleSendMessageSuccess)
        .recoverSuspend { error -> handleSendMessageFail(message, error) }
}
```